docs/adr/0003_canonicalization_algorithm.md

```diff
@@ -0,0 +1,83 @@
+# Introduce `O(n)` Canonicalization Algorithm
+
+* Status: Proposed
+* Authors: @LLFourn, @evanlinjin
+* Date: 2024-12-01
+* Targeted modules: `bdk_chain`
+* Associated Tickets/PRs: Issue #1665, ~PR #1659~, PR #1670
+
+## Context and Problem Statement
+
+The Wizardsardine BDK audit uncovered the severity of the performance issues in the original canonicalization logic. The problem is especially severe for wallet histories with many unconfirmed and conflicting transactions. This can be a dDos vector if BDK is used in server-side applications. The time complexity of the original canonicalization logic is `O(n^2)`.
```

Comment:
⛏️ Should include a link to the audit report.
```suggestion
The [2024 Wizardsardine BDK code audit](https://gist.github.com/darosior/4aeb9512d7f1ac7666abc317d6f9453b) uncovered the severity of the performance issues in the original canonicalization logic. The problem is especially severe for wallet histories with many unconfirmed and conflicting transactions. This can be a dDos vector if BDK is used in server-side applications. The time complexity of the original canonicalization logic is `O(n^2)`.
```